src/libraries/System.Private.CoreLib/src/System/IO/StreamReader.cs

```diff
@@ -883,63 +881,81 @@ private int ReadBuffer(Span<char> userBuffer, out bool readToUserBuffer)
-                int tmpCharLen = _charLen;
-                int tmpCharPos = _charPos;
-                int i = tmpCharPos;
+                return null;
+            }
 
+            char[]? rentedArray = null;
+            int lastWrittenIndex = 0;
+            try
+            {
                 do
                 {
-                    char ch = tmpCharBuffer[i];
-
                     // Note the following common line feed chars:
-                    // \n - UNIX   \r\n - DOS   \r - Mac
-                    if (ch == '\r' || ch == '\n')
+                    // \n - UNIX/Mac   \r\n - Windows
+                    int i = _charBuffer.AsSpan(_charPos, _charLen - _charPos).IndexOfAny('\r', '\n');
+                    if (i >= 0)
                     {
-                        string s;
-
-                        if (sb != null)
+                        string? s;
+                        if (rentedArray != null)
                         {
-                            sb.Append(tmpCharBuffer, tmpCharPos, i - tmpCharPos);
-                            s = sb.ToString();
+                            Append(ref rentedArray, lastWrittenIndex, _charBuffer, _charPos, i);
+                            lastWrittenIndex += i;
+                            s = new string(rentedArray!, 0, lastWrittenIndex);
                         }
                         else
                         {
-                            s = new string(tmpCharBuffer, tmpCharPos, i - tmpCharPos);
+                            s = new string(_charBuffer, _charPos, i);
                         }
-
-                        _charPos = tmpCharPos = i + 1;
-
-                        if (ch == '\r' && (tmpCharPos < tmpCharLen || (await ReadBufferAsync(cancellationToken).ConfigureAwait(false)) > 0))
+                        char ch = _charBuffer[_charPos + i];
+                        _charPos += i + 1;
+                        if (ch == '\r' && (_charPos < _charLen || (await ReadBufferAsync(cancellationToken).ConfigureAwait(false)) > 0))
                         {
-                            tmpCharPos = _charPos;
-                            if (_charBuffer[tmpCharPos] == '\n')
+                            if (_charBuffer[_charPos] == '\n')
                             {
-                                _charPos = ++tmpCharPos;
+                                _charPos++;
                             }
                         }
 
                         return s;
                     }
 
-                    i++;
-                } while (i < tmpCharLen);
-
-                i = tmpCharLen - tmpCharPos;
-                sb ??= new StringBuilder(i + 80);
-                sb.Append(tmpCharBuffer, tmpCharPos, i);
-            } while (await ReadBufferAsync(cancellationToken).ConfigureAwait(false) > 0);
+                    i = _charLen - _charPos;
+                    if (i < 0) break; // a hack for System.Globalization.Tests (to check on CI if it fixes the problem)
```

Comment:
can this be removed?

Reply:
yeah, sure. didn't notice it. will remove it shortly.

Reply:
Removed